compiler/rustc_target/src/spec/tests/tests_impl.rs

```diff
@@ -109,20 +112,56 @@ impl Target {
             );
         }
 
-        assert!(
-            (self.pre_link_objects_self_contained.is_empty()
-                && self.post_link_objects_self_contained.is_empty())
-                || self.link_self_contained != LinkSelfContainedDefault::False
-        );
+        if self.link_self_contained == LinkSelfContainedDefault::False {
+            assert!(
+                self.pre_link_objects_self_contained.is_empty()
+                    && self.post_link_objects_self_contained.is_empty()
+            );
+        }
 
         // If your target really needs to deviate from the rules below,
         // except it and document the reasons.
         // Keep the default "unknown" vendor instead.
         assert_ne!(self.vendor, "");
+        assert_ne!(self.os, "");
         if !self.can_use_os_unknown() {
             // Keep the default "none" for bare metal targets instead.
             assert_ne!(self.os, "unknown");
         }
+
+        // Check dynamic linking stuff
+        // BPF: when targeting user space vms (like rbpf), those can load dynamic libraries.
+        if self.os == "none" && self.arch != "bpf" {
+            assert!(!self.dynamic_linking);
+        }
+        if self.only_cdylib
+            || self.crt_static_allows_dylibs
+            || !self.late_link_args_dynamic.is_empty()
+        {
+            assert!(self.dynamic_linking);
+        }
+        // Apparently PIC was slow on wasm at some point, see comments in wasm_base.rs
+        if self.dynamic_linking && !(self.is_like_wasm && self.os != "emscripten") {
+            assert_eq!(self.relocation_model, RelocModel::Pic);
+        }
+        // PIEs are supported but not enabled by default with linuxkernel target.
+        if self.position_independent_executables && !triple.ends_with("-linuxkernel") {
+            assert_eq!(self.relocation_model, RelocModel::Pic);
+        }
+        if self.relocation_model == RelocModel::Pic {
+            assert!(self.dynamic_linking || self.position_independent_executables);
+        }
```

Comment:
This requirement is a bit odd since our default values for target options violate it -- the default relocation model is PIC, but the default for `dynamic_linking` and `position_independent_executables` is `false`. This is causing some trouble in https://github.com/rust-lang/rust/pull/133409.